### PR TITLE
Backport bloom filter improvements

### PR DESCRIPTION
**Pull request by str4d**
Cherry-picked from the following upstream PRs:

- bitcoin/bitcoin#7113
- bitcoin/bitcoin#7818
  - Only the second commit (to resolve conflicts).
- bitcoin/bitcoin#7934
- bitcoin/bitcoin#8655
  - Partial backport to help resolve conflicts.
- bitcoin/bitcoin#9060
- bitcoin/bitcoin#9223
- bitcoin/bitcoin#9644
  - Partial backport to help resolve conflicts.
- bitcoin/bitcoin#9916
- bitcoin/bitcoin#9750
- bitcoin/bitcoin#13176
- bitcoin/bitcoin#13948
- bitcoin/bitcoin#16073
- bitcoin/bitcoin#18670
- bitcoin/bitcoin#18806
  - Reveals upstream's covert fix for CVE-2013-5700.
- bitcoin/bitcoin#19968

### CODE REVIEW

**Comment by str4d** [comment]
@zkbot try

File: src/init.cpp [hunk 1]
```diff
@@ -551,14 +550,14 @@ struct CImportingNow
...
+    std::map<std::string, path> mapBlockFiles;
 
     // Glob all blk?????.dat and rev?????.dat files from the blocks directory.
     // Remove the rev files immediately and insert the blk file paths into an
     // ordered map keyed by block file index.
     LogPrintf("Removing unusable blk?????.dat and rev?????.dat files for -reindex with -prune\n");
-    path blocksdir = GetDataDir() / "blocks";
-    for (directory_iterator it(blocksdir); it != directory_iterator(); it++) {
+    boost::filesystem::path blocksdir = GetDataDir() / "blocks";
+    for (boost::filesystem::directory_iterator it(blocksdir); it != boost::filesystem::directory_iterator(); it++) {
```
Comment (str4d) [inline src/init.cpp:560]: Oops, this was a backport mistake (non-conflicting merge conflict due to me separately backporting the `boost::filesystem` refactor in #4805). I'll force-push a fix to the commit once zkbot is done testing the PR.

**Review by nuttycom — APPROVED** [review]
utACK

**Comment by str4d** [comment]
@zkbot r+

**Pull request merged**